solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -0,0 +1,369 @@
+#![allow(dead_code)]
+
+use super::single_pair_settlement::{AmmSwapExactTokensForTokens, SinglePairSettlement};
+use crate::settlement::Trade;
+use anyhow::{anyhow, Result};
+use model::order::{OrderCreation, OrderKind};
+use num::{bigint::Sign, BigInt};
+use std::collections::HashMap;
+use web3::types::{Address, U256};
+
+#[derive(Default)]
+struct TokenContext {
+    address: Address,
+    reserve: BigInt,
+    buy_volume: BigInt,
+    sell_volume: BigInt,
+}
+
+impl TokenContext {
+    pub fn is_excess(&self, other: &TokenContext) -> bool {
+        &self.reserve * (&other.sell_volume - &other.buy_volume)
+            < &other.reserve * (&self.sell_volume - &self.buy_volume)
+    }
+}
+
+/**
+ * Computes a settlement using orders of a single pair and the direct AMM between those tokens.
+ * Returns an error if computation fails, orders are not already filtered for a specific token pair, or the reserve information for that
+ * pair is not available.
+ */
+pub fn solve(
+    orders: impl Iterator<Item = OrderCreation> + Clone,
+    reserves: &HashMap<Address, U256>,
+) -> Result<SinglePairSettlement> {
```

Comment:
I'd rather have this not return a result and instead panic on error because the errors (more than one pair in the orders, internal bigint conversion fails, reserve missing) are programmer errors that should never happen.
Could change the types to encode that directly. Instead of Iterator Order we could take already split
sell_orders, buy_orders, sell_reserve, buy_reserve which removes the reserve hash map too.
The one error that we cannot prevent is if we get a bigint that is larger than u256 max but in that case we should just return None instead of an error.

Reply:
Agree that the programming errors should be turned into assertions. One logic error that I'm not sure how to handle is if our shortage is larger than the reserve inside one of the Uniswap pools. In this case you suggest to also return None?

Reply:
I think you're referring to

> We should better handle the case if the desired trade amount exceeds the reserve in the Uniswap pool (right now this would result in a bigint -> u256 conversion error.

but I don't understand this. This function doesn't know about the real state of the uniswap pool, it just creates a generic amm interaction. So where does that conversion error happen?

Reply:
Yes I would also expect None in that case since this set of orders cannot be solved.